### DISCUSSION TITLE
Spacepy for radiation belts modeling

### DISCUSSION THREAD

**Discussion by xandrd** [spacepy/spacepy#703] [Design and roadmapping]
Following advice from Jon Niehof, I open this issue.

I'm working with radiation belt modeling (in quasi-linear approximation) quite a lot and, personally, I could not find many cases of how spacepy can help me. Internally, my colleagues and I are using a library we developed in Matlab that does help us with many modeling tasks, like adiabatic invariants calculations and transformation, the calculation of drift velocities, getting output from some empirical models (e.g., plasmapause location, magnetopause location, radial diffusion coefficients etc.). Our library is quite helpful but it is not documented and does not have Python analog. I have submitted a PyHC proposal to develop such a library.

What I would like to discuss here is:
- the potential integration of such functionality
- general opinion from the community
- opinion from the spacepy developers on this subject. 
 
If the proposal is selected, the library will be lightweight and can be easily added as a dependence or installed individually. As for the use of spacepy, I would be interested to know in what cases spacepy helps with radiation belt modeling. 

Thanks!

**Comment by jtniehof** [comment]
Hi Sasha! I converted this to a discussion in our Design and Roadmapping section so we can kick this around a bit.

This is definitely within the scope of what we do in SpacePy and there are some bits of existing functionality. For instance, SpacePy has empirical plasmapause models, and of course the irbem wrapping offers a lot. There is also a simple 1D radbelt model and some data assimilation work--these aren't fully documented since they've been unmaintained for awhile, and are mostly suitable as learning tools. More functional research and operational codes would be very useful, and having them side-by-side with simpler models would be particularly beneficial.

I'd envision this as two categories of improvements: work on existing features (and some new features) roughly within the current scope, and chunks of new functionality. So please do look through what we've got and comment on what would make it more helpful, in addition to the brand-new.

I've had a contributor guide on my todo list for awhile, and that needs to happen, which would help with all this. The short list of things we'd be looking for:

- Relatively low-level in the sense of being able to use from a variety of applications / analysis code rather than forcing users into a particular framework (this sounds like your plan!)
- Sphinx-compatible docstrings
- Test coverage
- Integration with other features within spacepy where applicable, e.g. the datamodel, ticktock time objects, coordinates objects. This doesn't necessarily mean *only* using those, just that if there is support for a particular concept within SpacePy new code should try to support it.
